### PR TITLE
Issues/3506 page list ui

### PR DESCRIPTION
**Pull request by aerych**
Refs #3506 
This PR is the UI refresh of the Page list, and build on the work done in #3577 
With this change the page list gains support for:
- Filtering by status
- Local search
- Restoring a trashed page

### CODE REVIEW

**Comment by aerych** [comment]
@jleandroperez since you're already familiar with the post list changes would you be up for another review here?

File: WordPress/Classes/Categories/NSDate+StringFormatting.m [hunk 1]
```diff
@@ -46,4 +47,362 @@ - (NSString *)shortString
...
+                return  NSLocalizedString(@"Tomorrow", @"Tomorrow");
+            } else {
+                return  NSLocalizedString(@"Yesterday", @"Yesterday");
+            }
+        case 0:
+            if (isFutureDate) {
+                return  NSLocalizedString(@"Later today", @"Later today");
+            } else {
+                return  NSLocalizedString(@"Today", @"Today");
+            }
```
Comment (jleandroperez) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: Can we, perhaps, map the index into several segments?. That way, we would be able to use two Maps, this way:

```
NSDictionary *futureDateMap = @{
    @(0) : @"Later today",   // << Should be NSLocalizedString's
    @(1) : @"Tomorrow",
    @(2) : @"In %d days",
    @(3) : @"In a month",
    @(4) : @"In %d months",
    @(5) : @"In a year",
    @(6) : @"In %d years",
    @(7) : @"In many years"
};

NSDictionary *pastDateMap = @{
    // (..)
};

return isFutureDate ? futureDateMap[@(index)] : pastDateMap[@(index)]
```

(Plus `[NSString stringWithFormat: (...)`)

File: WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+@property (nonatomic, strong) IBOutlet NSLayoutConstraint *titleLabelTopMarginConstraint;
+@property (nonatomic, strong) IBOutlet NSLayoutConstraint *titleLabelBottomMarginConstraint;
+
+@end
+
+@implementation PageListTableViewCell
+
+@synthesize delegate;
+
+- (void)awakeFromNib {
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m]: Me, being very annoying: brace in the wrong spot :dancer: 

(sorry!)

File: WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    CGFloat availableWidth = innerWidth - CGRectGetWidth(self.menuButton.frame);
+    CGSize availableSize = CGSizeMake(availableWidth, CGFLOAT_MAX);
+
+    // Add up all the things.
+    CGFloat height = self.titleLabelTopMarginConstraint.constant;
+    height += self.titleLabelBottomMarginConstraint.constant;
+    height += [self.titleLabel sizeThatFits:availableSize].height;
+
+    height = MAX(height, CGRectGetHeight(self.menuButton.frame));
+    height += 1; // 1 pixel rule.
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m]: Been there!

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+
+    NSError *error;
+    Page *page = (Page *)[self.managedObjectContext existingObjectWithID:self.currentActionedPageObjectID error:&error];
+    if (error) {
+        DDLogError(@"%@, %@, %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd), error);
+    }
+
+    PostListStatusFilter filter = [self currentPostListFilter].filterType;
+
+    if (buttonIndex == 0) {
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: (Maybe constans?)

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+const CGFloat SearchWrapperViewPortraitHeight = 64.0;
+const CGFloat SearchWrapperViewLandscapeHeight = 44.0;
+
+@implementation AbstractPostListViewController
+
+#pragma mark - Lifecycle Methods
+
+- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
+{
+
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m]: (Maybe) `NSAssert(false, @"Please override me. Or crash. Mwah ha ha");` ?

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+            [self promptForPasswordWithMessage:nil];
+            return;
+        }
+    }
+    [WPError showNetworkingAlertWithError:error title:NSLocalizedString(@"Unable to Sync", @"Title of error prompt shown when a sync the user initiated fails.")];
+}
+
+- (void)promptForPasswordWithMessage:(NSString *)message
+{
+    if (message == nil) {
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m]: Is it possible there'd be any flow that would call this method with a nil message?

(If so maybe the 'default message' should be placed there)

File: WordPress/Classes/Categories/NSDate+StringFormatting.m (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: Great idea!

File: WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m (hunk 2 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListTableViewCell.m]: Oopsie! :)

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m (hunk 5 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m]: Arg! Missed one.  And its in the wrong place.

File: WordPress/Classes/Categories/NSDate+StringFormatting.m (hunk 1 above)
Comment (aerych) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: >  (Plus [NSString stringWithFormat: (...))

I don't think we can use `stringWithFormat` due to different languages having different (multiple) plural forms for numbers.

Comment (koke) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: I think it was @SergioEstevao who pointed me to [TTTTimeIntervalFormatter](https://github.com/mattt/FormatterKit#ttttimeintervalformatter) which already does this, handles plurals properly, and it's already translated.

**Comment by koke** [comment]
I tried to run this but it crashed:

```
2015-05-15 11:39:29.100 WordPress[84674:8891249] *** Terminating app due to uncaught exception 'NSUnknownKeyException', reason: '[<PageListViewController 0x7fe0a3c638b0> setValue:forUndefinedKey:]: this class is not key value coding-compliant for the key addButton.'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001148b0c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000114549bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x00000001148b08a9 -[NSException raise] + 9
	3   Foundation                          0x00000001111e6b53 -[NSObject(NSKeyValueCoding) setValue:forKey:] + 259
	4   CoreFoundation                      0x00000001147f8d50 -[NSArray makeObjectsPerformSelector:] + 224
	5   UIKit                               0x000000011264f52b -[UINib instantiateWithOwner:options:] + 1506
	6   UIKit                               0x00000001128b1aba -[UIStoryboard instantiateViewControllerWithIdentifier:] + 181
	7   WordPress                           0x000000010e50f1cd +[PageListViewController controllerWithBlog:] + 173
	8   WordPress                           0x000000010e473e5d -[BlogDetailsViewController showPageList] + 125
	9   WordPress                           0x000000010e4730b2 -[BlogDetailsViewController tableView:didSelectRowAtIndexPath:] + 786
	10  UIKit                               0x0000000112468dc9 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 1293
	11  UIKit                               0x0000000112468f0a -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 219
	12  UIKit                               0x000000011239b62c _applyBlockToCFArrayCopiedToStack + 314
	13  UIKit                               0x000000011239b4a6 _afterCACommitHandler + 533
	14  CoreFoundation                      0x00000001147e3ca7 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	15  CoreFoundation                      0x00000001147e3c00 __CFRunLoopDoObservers + 368
	16  CoreFoundation                      0x00000001147d9a33 __CFRunLoopRun + 1123
	17  CoreFoundation                      0x00000001147d9366 CFRunLoopRunSpecific + 470
	18  GraphicsServices                    0x000000011613aa3e GSEventRunModal + 161
	19  UIKit                               0x0000000112377900 UIApplicationMain + 1282
	20  WordPress                           0x000000010e3846ae main + 62
	21  libdyld.dylib                       0x0000000115453145 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

File: WordPress/Classes/Categories/NSDate+StringFormatting.m (hunk 1 above)
Comment (SergioEstevao) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: As an extra note this change (changing the format of Dates intervals) should de done in a separate PR. It's not related to the Pages redesign itself and I suppose it will affect the Posts section also.

Comment (aerych) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: >  TTTTimeIntervalFormatter

I'll take a look. Thanks for the tip!

> this change (changing the format of Dates intervals) should de done in a separate PR. It's not related to the Pages redesign itself

This formatting *is* specific to the pages list and not used elsewhere.

Comment (SergioEstevao) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: Don't you want to use it on the Post lists also? instead of 170d?

Comment (SergioEstevao) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: The Reader could also benefit if this new date formatters get in.

Comment (aerych) [inline WordPress/Classes/Categories/NSDate+StringFormatting.m]: > Don't you want to use it on the Post lists also? instead of 170d?

No. The page list uses the dates as section headers, not as a date in a card view. The page cells themselves do not display a date. So, its a different use case.

> The Reader could also benefit if this new date formatters get in.

Maybe. Probably. But that is a separate issue. :)

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m (hunk 6 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.m]: Actually, it looks like at this point every flow calls this with a nil message. Yay for old code?  I'll clean it up.

**Comment by aerych** [comment]
@koke Did this crash on launch? Maybe clean / ded to get rid of cached xibs?

**Comment by aerych** [comment]
@koke It should work now

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 4 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: A bit torn about this.  Since we're dealing with indexes, and since button index is situational I'm not sure creating consts would make this much clearer...

**Comment by aerych** [comment]
@jleandroperez Ready for round two! :)

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+        case PostListStatusFilterScheduled:
+            title = NSLocalizedString(@"You don't have any scheduled pages.", @"Displayed when the user views scheduled pages in the pages list and there are no pages");
+            break;
+        case PostListStatusFilterTrashed:
+            title = NSLocalizedString(@"You don't have any pages in your trash folder.", @"Displayed when the user views trashed in the pages list and there are no pages");
+            break;
+        default:
+            title = NSLocalizedString(@"You haven't published any pages yet.", @"Displayed when the user views published pages in the pages list and there are no pages");
+            break;
+    }
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Absolutely no need to change a single character here: This looks aaaaawesome with an inference map (and in Swift, it's even cooler!).

```
NSDictionary *filterMap = @{
   @(PostListStatusFilterDraft)  :  NSLocalizedString(@"You don't have any drafts.", nil),
   @(PostListStatusFilterScheduled) : //...
};

NSString *defaultTitle = NSLocalizedString(@"You haven't published (...)", nil);
return filterMap[@(filter.filterType)] ?: defaultTitle;
```

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m [hunk 8]
```diff
@@ -0,0 +1,531 @@
...
+    return CGFLOAT_MIN;
+}
+
+- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
+{
+    id <NSFetchedResultsSectionInfo> sectionInfo = [self.tableViewHandler.resultsController.sections objectAtIndex:section];
+    NSString *nibName = NSStringFromClass([PageListSectionHeaderView class]);
+    PageListSectionHeaderView *headerView = [[[NSBundle mainBundle] loadNibNamed:nibName owner:nil options:nil] firstObject];
+//    NSInteger index = [sectionInfo.name integerValue];
+//    NSString *title = [NSDate conciseStringFromIndex:index];
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Maybe nuke this guys?

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+{
+    [self configureCell:self.cellForLayout atIndexPath:indexPath];
+    CGSize size = [self.cellForLayout sizeThatFits:CGSizeMake(width, CGFLOAT_MAX)];
+    CGFloat height = ceil(size.height);
+    return height;
+}
+
+- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
+{
+    return 24.0;
```
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: (Maybe) constant?

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 4 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Sounds good!

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 7 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: one day... we will Swift all the things... ;)

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 8 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: doH!  Yes, indeed.

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 9 above)
Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Si.

**Comment by jleandroperez** [comment]
@aerych code looks awesome!.

#### Page List Not Refreshed:
1. Launch the app.
2. Tap over My Sites > (pick a testing site) > Pages.
3. Tap over the cell's right button, and tap over `Move to Draft`.

As a result, the post remained onscreen, even though the Titlebar filter was set to Publish.
Please note that **i triggered this just a couple times**, and it's not 10/10.

#### Crash:
I crashed the app, while playing with the steps outlined above. Console logs below.
**NSFetchedResultsController**'s delegate is set to *assign*. This would look like a FRC that remained alive a few extra milliseconds, and ended up hitting an invalid pointer.

Can we nuke the FRC's delegate in **WPTableViewHandler**'s dealloc?

2015-05-28 11:28:14.014 WordPress[1399:197835] -[OS_dispatch_queue controllerWillChangeContent:]: unrecognized selector sent to instance 0x174115720
2015-05-28 11:28:14.015 WordPress[1399:197835] CoreData: error: Serious application error.  Exception was caught during Core Data change processing.  This is usually a bug within an observer of NSManagedObjectContextObjectsDidChangeNotification.  -[OS_dispatch_queue controllerWillChangeContent:]: unrecognized selector sent to instance 0x174115720 with userInfo (null)
2015-05-28 11:28:14.016 WordPress[1399:197835] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[OS_dispatch_queue controllerWillChangeContent:]: unrecognized selector sent to instance 0x174115720'
*** First throw call stack:
(0x1821f82d8 0x193a1c0e4 0x1821ff3a4 0x1821fc154 0x1820feccc 0x181fbe530 0x181f3d218 0x181f3d0e4 0x181ec8650 0x18219e2c4 0x1820db450 0x18300aa80 0x181ec85ac 0x181ec852c 0x181ec69d4 0x181ec5794 0x10003120c 0x100030230 0x181f3d218 0x101720f94 0x101725c28 0x1821af7f8 0x1821ad8a0 0x1820d92d4 0x18b8ef6fc 0x186c9efac 0x1000330e8 0x19409aa08)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb)

File: WordPress/Classes/ViewRelated/Pages/PageListViewController.m (hunk 7 above)
Comment (koke) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Unless `filter == nil` and this :boom:s  :grin:

Comment (aerych) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: `currentPostListFilter` should always return a filter

Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Pages/PageListViewController.m]: Plus, even if `filter == nil`, nothing would break. Worst case scenario, you can always handle defaults.

```
PostListStatusFilter filter = filter ? filter.filterType : PostListStatusFilterDefault
```

**Comment by aerych** [comment]
> Please note that i triggered this just a couple times, and it's not 10/10.

Thanks @jleandroperez.  So far I haven't been able to reproduce the issue with drafts, but I have a theory.  I neglected to show a prompt if there was an error, so maybe there was a problem setting the post to a draft that failed silently.  If it happened quick enough you wouldn't see the UI change at all.  I've added a prompt with the latest commit.  Could you try to repo again? 

I also set the FRC delegate to nil in dealloc so hopefully that takes care of the crash.

**Comment by jleandroperez** [comment]
@aerych couldn't trigger the crash (OR) the UI glitch anymore. I'm pretty sure that the TableViewHandler patch will take care of the scenario that triggered the crash, so we're good!.

Awesome job Eric!

:shipit:

**Comment by aerych** [comment]
Thanks @jleandroperez !
